### PR TITLE
Allow multiple providers for http client

### PR DESCRIPTION
**Pull request by dharjeezy**
Introduces a Chained middleware to allow for possible multiple providers for request

### CODE REVIEW

File: ethers-providers/src/rpc/transports/middleware.rs [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+                log::trace!(target:"ethers-providers", "Possibly encountered an os error submitting request, switching provider {e:?}");
+                let _ = next_state(None)?;
+            }
+        }
+
+        Ok(None)
+    }
+
+    fn max_chain_length(&self) -> u32 {
+        MAX_CHAIN_LENGTH
```
Comment (Wizdave97) [inline ethers-providers/src/rpc/transports/middleware.rs]: We can set max chain length to the number of providers + 1

File: ethers-providers/src/rpc/transports/middleware.rs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+                if response.status() != StatusCode::OK {
+                    match response.error_for_status_ref() {
+                        Ok(_res) => (),
+                        Err(err) => {
+                            let _ = next_state(Some(ReqwestError(err)))?;
+                        }
+                    }
+                };
+
+                let maybe_body = response.chunk().await?;
```
Comment (Wizdave97) [inline ethers-providers/src/rpc/transports/middleware.rs]: Does this method return the full request body in one call?

File: ethers-providers/src/rpc/transports/middleware.rs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+            Ok::<_, anyhow::Error>(())
+        };
+
+        match result {
+            Ok(mut response) => {
+                if response.status() != StatusCode::OK {
+                    match response.error_for_status_ref() {
+                        Ok(_res) => (),
+                        Err(err) => {
+                            let _ = next_state(Some(ReqwestError(err)))?;
```
Comment (Wizdave97) [inline ethers-providers/src/rpc/transports/middleware.rs]: Doesn't this line need a return statement?
```suggestion
                            let _ = next_state(Some(ReqwestError(err)))?;
```

File: ethers-providers/src/rpc/transports/middleware.rs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+                        }
+                    }
+                };
+
+                let maybe_body = response.chunk().await?;
+
+                if let Some(body) = maybe_body {
+                    if !String::from_utf8_lossy(body.as_ref()).contains("jsonrpc") {
```
Comment (Wizdave97) [inline ethers-providers/src/rpc/transports/middleware.rs]: Don't think this is required, the json deserialization takes care of this case

**Review by Wizdave97 — CHANGES_REQUESTED** [review]

File: ethers-providers/src/rpc/transports/middleware.rs (hunk 4 above)
Comment (dharjeezy) [reply ethers-providers/src/rpc/transports/middleware.rs]: When i tested this, the http call(that is not rpc) was failing as the deserialization didn't take care if it. I was having a Serde deserialization error

File: ethers-providers/src/rpc/transports/middleware.rs (hunk 2 above)
Comment (dharjeezy) [reply ethers-providers/src/rpc/transports/middleware.rs]: i have updated this implementation

File: ethers-providers/src/rpc/transports/middleware.rs (hunk 3 above)
Comment (dharjeezy) [reply ethers-providers/src/rpc/transports/middleware.rs:81]: done now

File: ethers-providers/src/rpc/transports/middleware.rs (hunk 4 above)
Comment (Wizdave97) [reply ethers-providers/src/rpc/transports/middleware.rs]: All json rpc responses contain that string, even the error responses. That check makes everything after it useless

Comment (dharjeezy) [reply ethers-providers/src/rpc/transports/middleware.rs]: Yeah right. The check is actually for response bodies that are not json rpc related.  It is an inverse check.

Comment (Wizdave97) [reply ethers-providers/src/rpc/transports/middleware.rs]: That's why I'm saying the deserialization will catch that, it doesn't deserialize it's an error response.

Comment (Wizdave97) [reply ethers-providers/src/rpc/transports/middleware.rs]: We can't act on a response that it not json rpc formatted

**Comment by Wizdave97** [comment]
Also you should integrate the middleware into the http provider

File: ethers-providers/src/rpc/transports/http.rs [hunk 5]
```diff
@@ -193,21 +194,20 @@ impl Provider {
     /// let provider = Http::new_with_client(url, client);
     /// ```
     pub fn new_with_client(url: impl Into<Url>, client: reqwest::Client) -> Self {
-        Self { id: AtomicU64::new(1), client, url: url.into() }
+        Self { id: AtomicU64::new(1), client, url: url.into(), retry_urls: vec![] }
```
Comment (Wizdave97) [inline ethers-providers/src/rpc/transports/http.rs]: Put single url inside the vector

File: ethers-providers/src/rpc/transports/http.rs [hunk 6]
```diff
@@ -254,7 +254,12 @@ impl FromStr for Provider {
 
 impl Clone for Provider {
     fn clone(&self) -> Self {
-        Self { id: AtomicU64::new(1), client: self.client.clone(), url: self.url.clone() }
+        Self {
+            id: AtomicU64::new(1),
+            client: self.client.clone(),
+            url: self.url.clone(),
+            retry_urls: vec![],
```
Comment (Wizdave97) [inline ethers-providers/src/rpc/transports/http.rs]: why is this not cloned

File: ethers-providers/src/rpc/transports/http.rs [hunk 7]
```diff
@@ -193,21 +194,20 @@ impl Provider {
     /// let provider = Http::new_with_client(url, client);
     /// ```
     pub fn new_with_client(url: impl Into<Url>, client: reqwest::Client) -> Self {
-        Self { id: AtomicU64::new(1), client, url: url.into() }
+        Self { id: AtomicU64::new(1), client, url: url.into(), retry_urls: vec![] }
     }
 
     /// Allows you to make a request with multiple providers
     pub async fn request_with_chain_middleware<T: Serialize + Send + Sync, R: DeserializeOwned>(
```
Comment (Wizdave97) [inline ethers-providers/src/rpc/transports/http.rs]: Just have a method that creates the provider with the middleware `new_client_with_chain_middleware` and all provider urls, that's really all there needs to be, nothing else needs to change in the provider

**Review by Wizdave97 — CHANGES_REQUESTED** [review]

File: ethers-providers/src/rpc/transports/http.rs [hunk 8]
```diff
@@ -186,7 +194,52 @@ impl Provider {
...
+        let client = ClientBuilder::new(self.client.clone())
+            .with(ChainMiddleware::new(SwitchProviderMiddleware::_new(self.retry_urls.clone())))
+            .build();
+
+        let res = client
+            .post(self.url.as_ref())
+            .json(&payload)
+            .send()
+            .await
+            .map_err(|err| ClientError::MiddlewareError(err))?;
```
Comment (Wizdave97) [inline ethers-providers/src/rpc/transports/http.rs]: This is duplicated code.

**Review by Wizdave97 — CHANGES_REQUESTED** [review]

File: ethers-providers/src/rpc/transports/middleware.rs [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+                    .any(|_| true);
+
+                if res {
+                    return Err(anyhow!("All providers returned {:?}", StatusCode::NOT_FOUND))?;
+                }
+                next_index = 0;
+            }
+            _state.active_provider_index = next_index;
+            let next_provider = self.providers[next_index].clone();
+            let url_ref = request.url_mut();
```
Comment (Wizdave97) [inline ethers-providers/src/rpc/transports/middleware.rs:55]: Have you inspected that the payload is maintained when the urls are switched, there should be a test for that.

**Review by Wizdave97 — CHANGES_REQUESTED** [review]

File: ethers-providers/src/rpc/transports/middleware.rs [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+        &self,
+        result: Result<reqwest::Response, Error>,
+        _state: &mut Self::State,
+        request: &mut reqwest::Request,
+    ) -> Result<Option<reqwest::Response>, Error> {
+        let mut next_state = |client_error: Option<ClientError>| {
+            let active_index = _state.active_provider_index;
+            _state.prev_stat.insert(active_index, client_error);
+            let mut next_index = _state.active_provider_index + 1;
+            if next_index >= self.providers.len() {
```
Comment (Wizdave97) [inline ethers-providers/src/rpc/transports/middleware.rs:44]: If the next_index >= self.providers.len(), just exit the middleware chain

File: ethers-providers/src/rpc/transports/retry.rs [hunk 11]
```diff
@@ -409,6 +409,7 @@ impl RetryPolicy<ClientError> for HttpRateLimitRetryPolicy {
                 }
                 false
             }
+            _ => false,
```
Comment (Wizdave97) [inline ethers-providers/src/rpc/transports/retry.rs]: Don't use a catch all.

File: ethers-providers/src/rpc/transports/middleware.rs [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+/// Middleware for switching between providers on failures
+pub struct SwitchProviderMiddleware {
+    /// Providers for the url
+    pub providers: Vec<Url>,
+}
+
+#[derive(Default, Debug)]
+pub struct LocalState {
+    pub active_provider_index: usize,
+    pub prev_stat: HashMap<usize, Option<ClientError>>,
```
Comment (Wizdave97) [inline ethers-providers/src/rpc/transports/middleware.rs]: Don't think we need this anymore
```suggestion
```

File: ethers-providers/src/rpc/transports/middleware.rs [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+        result: Result<reqwest::Response, Error>,
+        _state: &mut Self::State,
+        request: &mut reqwest::Request,
+    ) -> Result<Option<reqwest::Response>, Error> {
+        let mut next_state = |client_error: Option<ClientError>| {
+            let active_index = _state.active_provider_index;
+            _state.prev_stat.insert(active_index, client_error);
+            let next_index = _state.active_provider_index + 1;
+            if next_index >= self.providers.len() {
+                return Err(anyhow!("Exiting Middleware"))?;
```
Comment (Wizdave97) [inline ethers-providers/src/rpc/transports/middleware.rs]: Let's return the client error that was passed if it's not None, or create one

Comment (Wizdave97) [reply ethers-providers/src/rpc/transports/middleware.rs]: You can add a trace log that lets the consumer know the providers have been exhausted.

**Review by Wizdave97 — CHANGES_REQUESTED** [review]

**Review by Wizdave97 — APPROVED** [review]

**Pull request merged**